lib/gzip_vas.c

```diff
@@ -319,28 +327,88 @@ int nxu_run_job(nx_gzip_crb_cpb_t *cmdp, nx_devp_t nxhandle)
 				return ret;
 			}
 		}
-		else {
-			/* paste has failed; should happen when NX
-			   queue is full or the paste buffer in the
-			   cache was being used
-			*/
-			/* If the NX queue was full, it should complete at
-			   least 2 requests after 50 timebase ticks.
-			   If the system is under very high load, we might see
-			   a couple of failures. In that case, it is a good
-			   idea to sleep this thread in order to give the other
-			   threads/processes CPU time to complete their
-			   execution. */
-			ticks_total = nx_wait_ticks(500, ticks_total, 0);
-
-			if (ticks_total > (timeout_seconds * nx_get_freq()))
-				return -ETIMEDOUT;
-
-			++retries;
-			if (retries % 1000 == 0) {
-				prt_err("Paste attempt %d, failed pid= %d\n", retries, getpid());
+
+		/* Paste has failed. Can happen for a few reasons:
+		   1. NX queue is full
+		   2. paste buffer in the cache was being used
+
+		   These apply exclusively to PowerVM:
+		   3. VAS window suspended after a DLPAR core remove op
+		   4. partition migration is in progress */
+		++retries;
+		if (retries % 1000 == 0) {
+			prt_err("Paste attempt %d, failed pid= %d\n", retries,
+				getpid());
+		}
+
+		/* On PowerVM, the failures may be caused by a suspended window.
+		   Try a few times and then check the number of credits. If the
+		   system is oversubscribed or the number of total credits
+		   reduced from the time the window was created, then our window
+		   may be suspended. There's no way to be sure, so be
+		   pessimistic and assume it is. */
+		if (nx_config.virtualization == POWERVM &&
+		    ticks_total > (freq >> 5) &&
```

Comment:
It's a good idea to move `(freq >> 5)` to a macro with a proper name, giving a better idea of what is going on, e.g. `SUSPENDED_TIME_THRESHOLD`.

Reply:
Done.